x/gamm/pool-models/balancer/amm.go

```diff
@@ -331,6 +333,53 @@ func getPoolAssetsByDenom(poolAssets []PoolAsset) (map[string]PoolAsset, error)
 	return poolAssetsByDenom, nil
 }
 
+// updateIntermediaryPoolAssets updates poolAssetsByDenom with liquidity.
+//
+// all liqidity coins must exist in poolAssetsByDenom. Returns error, if not.
+//
+// This is a helper function that is useful for updating the pool asset amounts
+// as an intermediary step in a multi-join methods such as CalcJoinPoolShares.
+// In CalcJoinPoolShares with multi-asset joins, we first attempt to do
+// a MaximalExactRatioJoin that might leave out some tokens in.
+// Then, for every remaining tokens in, we attempt to do a single asset join.
+// Since the first step (MaximalExactRatioJoin) affects the pool liqudity due to slippage,
+// we would like to account for that in the subsequent steps of single asset join.
+func updateIntermediaryPoolAssets(liquidity sdk.Coins, poolAssetsByDenom map[string]PoolAsset) error {
+	for _, coin := range liquidity {
+		poolAsset, ok := poolAssetsByDenom[coin.Denom]
+		if !ok {
+			return fmt.Errorf(errMsgFormatFailedInterimLiquidityUpdate, coin.Denom)
+		}
+
+		poolAsset.Token.Amount = poolAssetsByDenom[coin.Denom].Token.Amount.Add(coin.Amount)
+		poolAssetsByDenom[coin.Denom] = poolAsset
+	}
+	return nil
+}
+
+// calcJoinSingleAssetTokensIn attempts to calculate single
+// asset join for all tokensIn given totalShares in pool,
+// poolAssetsByDenom and swapFee. totalShares is the number
+// of shares in pool before beginnning to join any of the tokensIn.
+//
+// Returns totalNewShares and totalNewLiquidity from joining all tokensIn
+// by mimicking individually single asset joining each.
+// or error if fails to calculate join for any of the tokensIn.
+func (p *Pool) calcJoinSingleAssetTokensIn(tokensIn sdk.Coins, totalShares sdk.Int, poolAssetsByDenom map[string]PoolAsset, swapFee sdk.Dec) (sdk.Int, sdk.Coins, error) {
```

Comment:
Am I correct in saying that this function assumes `len(tokensIn)` < `len(poolAssets)` i.e. that the number of tokens in `tokensIn` is less than the total number of assets/denoms in the pool since otherwise, we would run `MaximalExactRatioJoin` instead of running `calcSingleAssetJoin` on each asset?

If so, it might be worth enforcing this with a check at the beginning of the function so that we only have one valid way of doing a multi-asset join: `MaximalExactRatioJoin` following by `calcJoinSingleAssetTokensIn` on the remaining assets

Reply:
While I agree that this check is useful, I think that we should have it in the caller of this function, not inside this function itself.

A function should work independently of its caller. Otherwise, we would add an invariant that might restrict its use in other cases.

I will investigate this further

Reply:
What do you think about having this check here: 
https://github.com/osmosis-labs/osmosis/commit/f6071c9757448dc929f3933b3ead336efcbcd9e3

Reply:
> A function should work independently of its caller. Otherwise, we would add an invariant that might restrict its use in other cases.

Generally agree with this - probably a good idea to leave it as it is to not break abstraction

> What do you think about having this check here: [f6071c9](https://github.com/osmosis-labs/osmosis/commit/f6071c9757448dc929f3933b3ead336efcbcd9e3)

This addresses a separate but important point & lets us have an error case we can use for testing. Great addition